types/clean-css/index.d.ts

```diff
@@ -650,12 +645,30 @@ declare namespace CleanCSS {
         minify(sources: Sources, sourceMap?: string): Promise<Output>;
     }
 
+    /**
+     * Options when returning a promise
+     */
+    type OptionsPromise = OptionsBase & { returnPromise: true };
+
+    /**
+     * Options when returning an output
+     */
+    type OptionsOutput = OptionsBase & { returnPromise?: false };
+
+    /**
+     * Discriminant union of both sets of options types.  If you initialize without setting `returnPromise: true`
+     *  and want to return a promise, you will need to cast to the correct options type so that TypeScript
+     *  knows what the expected return type will be:
+     *  `(options = options as CleanCSS.OptionsPromise).returnPromise = true`
+     */
+    type Options = OptionsPromise | OptionsOutput;
+
     /**
      * Constructor interface for CleanCSS
      */
     interface Constructor {
-        new(options: Options & { returnPromise: true }): MinifierPromise;
-        new(options?: Options): MinifierOutput;
+        new(options: OptionsPromise): MinifierPromise;
```

Comment:
Tested those changes and it fails the tests when dealing with trying to pass options as an object.

```javascript
let CleanCssOptions: CleanCSS.Options = { returnPromise: true };
new CleanCSS(CleanCssOptions).minify(source)  // <-- error: expect it to return MinifierPromise instead returns MinifierOutput
    .then((minified: CleanCSS.Output): void => {
        console.log(minified.styles);
    }).catch((error: any): void => {
        console.log(error);
    }
);
```

I tried an approach similar to this but TypeScript doesn't recognize `{ returnPromise: true }` without creating two separate Options types with a discriminated union.

Reply:
I will add, I think the reason that trying to check if the options extends `{ returnPromise: true }` doesn't work is that, in the original way method of doing the objects interface, it has the property `{ returnPromise: boolean }` and `{ returnPromise: true }` does not inherit from `{ returnPromise: boolean }`.  This means if I pass `{ returnPromise: true }` directly to the constructor, it works but if I create an object instance of options, it doesn't inherit `{ returnPromise: true }` but rather the `{ returnPromise: boolean }`-- if all that makes sense.  In this case `true` isn't a value per say but rather a type literal and `boolean` cannot be assigned to `true`.

Reply:
Last thing to add.  So the only thing that doesn't work in the code change I made is this scenario.

```javascript
let CleanCssOptions: CleanCSS.Options = { };
CleanCssOptions.returnPromise = true;   // <-- error as TypeScript assigned the options branch for the output and it only allows returnPromise = false
// to fix this we will need to manually re-cast the options
(CleanCssOptions = CleanCssOptions as CleanCSS.OptionsPromise).returnPromise = true;
// this works since we are switching the casting to use the promise options branch.
```

Other then this one scenario, by splitting options in OptiomsPromise and OptionsOutput with the discriminated union between them, we are able to property return the correct minifier with the correct return type regardless if we pass configuration properties direct or pass configuration properties using an object.